Add support for editing directories

omus opened the pull request:
I wanted to be able to more easily edit packages. By adding directory support to `edit` you can now do things like `edit(Pkg.dir("TimeZones"))` to open an entire directory in your preferred editor.

The change required me to optionally add in arguments for line numbers. This was required by sublime to work correctly and I think it makes sense not to include them for other editors as well. I'll note that editors such as nano and TextEdit don't support opening directories but fail gracefully.

omus commented:
I'll be updating the documentation

stevengj commented on base/interactiveutil.jl:
```diff
@@ -23,31 +23,35 @@ function editor()
     const no_line_msg = "Unknown editor: no line number information passed.\nThe method is defined at line $line."
     if startswith(name, "emacs") || name == "gedit"
-        spawn(`$command +$line $file`)
+        args = isfile(path) ? `+$line $path` : `$path`
```
I guess this works, but it seems cleaner to me to use an array here for `args`, i.e. `["+$line" path] : [path]`.

mbauman commented:
Huh, I've been doing exactly that for a long time.  I suppose the editors I've used (Atom and TextMate) just ignore line numbers for directories.  Would it be cleaner to only pass line numbers when they are given?  E.g., use a sentinel value for the default line number instead of checking isfile or isdir?

omus commented:
Interesting. Maybe this is only really an issue for Sublime. Running `subl .:1` causes sublime to make a new file that matches the name of the directory you are in. If you do `subl .` it will open the directory.

omus commented:
I switched the `isfile` checks to `line != 0`. I don't think any editors have line numbers starting at 0. Additionally the code no longer displays the line number warning when you aren't using them.

omus committed:
Refactored edit command execution

run/spawn command now occurs only in a single place making future
editing easier.

omus committed:
Excluding line info when line is zero

omus committed:
Rename edit argument from file to path

omus committed:
Update edit documentation to mention directories

omus commented:
I've updated the code to be up to date and did some refactoring to assist in future maintainability. @mbauman let me know what you think.

stevengj commented:
LGTM.

mbauman commented:
Agreed! This looks great.  Thanks @omus.

swt30 commented on base/docs/helpdb.jl:
```diff
@@ -1335,11 +1335,11 @@ Get the value type of an associative collection type. Behaves similarly to `elty
+Edit a file or directory optionally providing a line number to edit the file at. Returns to the julia prompt when you quit the editor.
 """
-edit(file::AbstractString, line=?)
+edit(path::AbstractString, line=?)
```
What does the question mark do in this context? I see it's in a few other places in helpdb but I've never seen this syntax before, and it's difficult to search for...

omus replied:
I believe it is just indicating that the line field is optional. http://julia.readthedocs.org/en/latest/stdlib/base/#Base.edit
